### ISSUE TITLE
Can you make one flow depend on another?

### ISSUE THREAD

**Issue by tchoedak** [azkaban/azkaban#741]
I have some scenarios where one flow could depend on another. A simple way to do it is to put all the jobs from the second flow just into the first flow and chain the dependencies. One issue with that is this could become a monolithic beast of a flow with many hands touching it if the flow becomes complex and serves multiple users. So I'm wondering if instead you can create a flow that depends on either another flow successfully completing all the way, or if you could make it depend on a single job or group of jobs from another flow finishing.

**Comment by logiclord** [comment]
No, at present you cannot have one flow depend on any other flow. 

The closes thing supported for this is to make multiple executions of a flow wait e.g. We have E1 already running for a Flow F1 in project P then another execution say E2 can be made to until E1 finishes. 

On a side note, it is very straight forward to add a feature and make any execution depend on any other execution by generalizing (asking user/config for some execution dependency information) and making changes at https://github.com/azkaban/azkaban/blob/master/azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java#L999 @HappyRay thoughts ?

**Comment by HappyRay** [comment]
Thanks @logiclord for answering questions.
This is an interesting use case. I haven't seen enough demand yet to make it a priority. The team is busy automating the build, deployment, making it easier to operate in a large environment, fixing bugs etc.

**Comment by tolgakonik** [comment]
+1 on this. In absence of this, subcomponents run at independent schedules, leading to either latency or serious compute waste.

waiting is not a great solution because that prevents you from defining serious SLAs.

**Comment by cdonat** [comment]
We'd be very interested in such a feature as well. Our scenario is, that we have one daily flow and a monthly one, that should only start after the daily one has finished for the 1st of that month.

**Comment by togoon** [comment]
+1, i'm also very looking forward to this function

**Comment by devangorder** [comment]
+1 - we have been struggling with the same issue as @cdonat for a while here

**Comment by wangqiaoshi** [comment]
+1,we need this function too!

**Comment by FrommyMind** [comment]
+1,we need this function too!

**Comment by ameyamk** [comment]
Planned and on roadmap. Likely will be prod ready in Q3 of 2017

**Comment by roshfsk** [comment]
+1, appreciate if you could provide us with an update on the status of this enhancement. It is a much needed feature.

**Comment by garyelephant** [comment]
+1

**Comment by peaksnail** [comment]
+1

**Comment by elonlo** [comment]
+1,we need this function too!

**Comment by vickyi** [comment]
+1, we need this function too!!

**Comment by DataScientistSamChan** [comment]
+1, we need this function too!!

**Comment by magicsilence** [comment]
+1, wo need this function

**Comment by toby-allsopp** [comment]
I also have this requirement. I will be looking into using the flow trigger feature with a dependency plugin that looks for successful runs of other flows. If it suits our purposes I'll submit a PR.

**Comment by burgerkingeater** [comment]
@toby-allsopp thanks, would you mind sharing your design first when ready? I'm interested in your solution. Kafka based dependency type +  flow/job event reporting(https://github.com/azkaban/azkaban/pull/1332) which we already have definitely can be the building block of this feature. But flow trigger needs to be extended as well to make it work with flow-to-flow dependency. I have something in mind and I'm happy to discuss it with you when you are ready to work on this.

**Comment by toby-allsopp** [comment]
@burgerkingeater thanks for your input!

So, it looks like the existing Kafka event reporting + the Kafka dependency type could be used to do something like what we need, at the expense of having to set up Kafka. What extensions are needed to the flow trigger feature to make this work?

If possible, I'd prefer to avoid a dependency on Kafka so I was imagining writing a new dependency plugin that is able to query the Azkaban execution history to look for successful (or possibly any completed) executions within a configurable time window. However, it doesn't look like dependency plugins have any access to the Azkaban back-end database or APIs. Is this going to be a fatal problem for this kind of approach?

**Comment by burgerkingeater** [comment]
@toby-allsopp currently flow trigger is time-based. When the time comes, a flow trigger instance will be created checking the availability of defined data dependencies.
E.g:

```
trigger:
  maxWaitMins: 60
  schedule:
     type: cron
     value: <*cron expression>
triggerDependencies:
 // some data dependencies.
```

However if we extend flow trigger to embrace streaming style - new flow trigger instance will be created when last flow trigger instance finishes, and at any given time there's only one flow trigger instance is checking the availability of defined data set, then it should achieve the purpose of flow-to-flow or job-to-flow dependency when the data set is kafka. 

The motivation of need for kafka is it opens possibility of flow-to-flow dependency across two different azkaban installations.

**Comment by toby-allsopp** [comment]
The time-based flow triggers are exactly what we want. To illustrate the requirement we are trying to solve, imagine two flows: "Daily" and "Weekly". "Daily" is scheduled to run every day at 8pm. "Weekly" is scheduled to run every Friday at 10pm, BUT we want to ensure that the "Daily" run on Friday has finished before "Weekly" starts (either due to resource constraints or data dependencies).

**Comment by burgerkingeater** [comment]
@toby-allsopp but it won't fit the use case where one flow depends another, right?

**Comment by toby-allsopp** [comment]
I'm not sure what that use case looks like - does it mean starting one flow as soon as another flow has finished? What does that look like if a flow depends on several other flows?

**Comment by burgerkingeater** [comment]
@toby-allsopp  
does it mean starting one flow as soon as another flow has finished?
-Yes
 What does that look like if a flow depends on several other flows?
- It won't be supported.

**Comment by toby-allsopp** [comment]
That sounds like a very different feature to what I'm thinking of, but it does sound useful, particularly with the executor tags feature I've proposed in #2154. With that combination you could have a workflow that combines several flows each with different execution environment requirements.

Do you think this could be thought of as a new job type that triggers the execution of a named flow?

**Comment by burgerkingeater** [comment]
I think it should fit into flow trigger, which is created to encapsulate
all different ways of triggering a flow.

**Comment by toby-allsopp** [comment]
> I think it should fit into flow trigger, which is created to encapsulate all different ways of triggering a flow.

@burgerkingeater Does this mean that eventually regular schedules will be replaced by or implemented in terms of flow triggers?

**Comment by davesiddons** [comment]
Hi, hope you dont mind me dropping some thoughts here.

Interflow dependency is a biggy for us and I've always been quite surprised at its absence in modern workflow tooling. Its a given in traditional job schedulers.

The things we do that would really benefit form this is around various ETL jobs and overnight batches we run that are all interdependent in some way. Currently the only way we can handle this is dropping flag files at the end of a flow to trigger another flow which is horrible and not very visible. It leads to all kinds of hidden issues when flows fail and don't housekeep correctly. What would be awesome would be the ability for a scheduled execution of a flow to be configured to run upon the completion of another scheduled execution of a different flow. This would ensure that flows only ran after the correct run of a preceding flow(s). It should be flexible enough to define whether or not the preceding flow(s) had to be completed successfully. So, just as you have dependencies between jobs you would have the same between flows. The care would be needed to ensure the dependency was based on a specific execution of a preceding flow. Extending this to having flows dependant on the completion of a job within another execution of a flow would be cool too.

This sort of introduces the idea of a calendar. So you schedule a flow daily at a given time. Then you schedule another flow daily at a later time but also with a dependency on completion of the same days 1st flow. This could introduce some quite complex calendar based rules but very useful.

Sorry for the brain dump and please ignore if already covered

**Comment by liangrui1988** [comment]
+1

**Comment by riritack** [comment]
+1

**Comment by zxxsinventory** [comment]
+1, we need this function too!!

**Comment by sansna** [comment]
+1, consider it plz

**Comment by basebase** [comment]
+1, consider it plz
